qa/tasks/cephfs/test_client_recovery.py

```diff
@@ -567,3 +567,37 @@ def test_abort_conn(self):
         self.assert_session_state(gid, "open")
         time.sleep(session_timeout * 1.5)  # Long enough for MDS to consider session stale
         self.assert_session_state(gid, "stale")
+
+    def test_dont_mark_unresponsive_client_stale(self):
+        """
+        Test that an unresponsive client holding caps not required by any
+        other client is evicted directly, without being marked as stale even
+        after session_timeout, at session_autoclose.
+        """
+        if not isinstance(self.mount_a, FuseMount):
+            self.skipTest("Require FUSE client to handle signal STOP/CONT")
+
+        # XXX: To conduct this test we need at least two clients since a
+        # single client is never evcited by MDS.
+        SESSION_TIMEOUT = 30
+        SESSION_AUTOCLOSE = 50
+        mount_a_gid = self.mount_a.get_global_id()
+        mount_a_pid = self.mount_a.get_client_pid()
+        self.fs.set_var('session_timeout', SESSION_TIMEOUT)
+        self.fs.set_var('session_autoclose', SESSION_AUTOCLOSE)
+        self.assert_session_count(2, self.fs.mds_asok(['session', 'ls']))
+
+        # test that client holding cap not required by any other client is not
+        # marked stale when it becomes unresponsive.
+        self.mount_a.run_shell(['mkdir', 'dir'])
+        self.mount_a.send_signal('sigstop')
+        time.sleep(SESSION_TIMEOUT + 2)
+        self.assert_session_state(mount_a_gid, "open")
+
+        # test that other clients have to wait to get the caps from
+        # unresponsive client until session_autoclose.
+        self.mount_b.run_shell(['stat', 'dir'])
+        time.sleep(SESSION_AUTOCLOSE - SESSION_TIMEOUT + 2)
+        self.assert_session_count(1, self.fs.mds_asok(['session', 'ls']))
+
+        self.mount_a.send_signal('sigcont')
```

Comment:
Okay so this call hangs:

````
2019-06-21T10:26:42.488 INFO:teuthology.orchestra.run:Running command with timeout 900
2019-06-21T10:26:42.488 INFO:teuthology.orchestra.run.smithi141:Running:
2019-06-21T10:26:42.489 INFO:teuthology.orchestra.run.smithi141:> sudo python2 -c '
2019-06-21T10:26:42.489 INFO:teuthology.orchestra.run.smithi141:> import glob
2019-06-21T10:26:42.489 INFO:teuthology.orchestra.run.smithi141:> import re
2019-06-21T10:26:42.489 INFO:teuthology.orchestra.run.smithi141:> import os
2019-06-21T10:26:42.489 INFO:teuthology.orchestra.run.smithi141:> import subprocess
2019-06-21T10:26:42.489 INFO:teuthology.orchestra.run.smithi141:>
2019-06-21T10:26:42.489 INFO:teuthology.orchestra.run.smithi141:> def find_socket(client_name):
2019-06-21T10:26:42.490 INFO:teuthology.orchestra.run.smithi141:>         asok_path = "/var/run/ceph/ceph-client.0.*.asok"
2019-06-21T10:26:42.490 INFO:teuthology.orchestra.run.smithi141:>         files = glob.glob(asok_path)
2019-06-21T10:26:42.490 INFO:teuthology.orchestra.run.smithi141:>
2019-06-21T10:26:42.490 INFO:teuthology.orchestra.run.smithi141:>         # Given a non-glob path, it better be there
2019-06-21T10:26:42.490 INFO:teuthology.orchestra.run.smithi141:>         if "*" not in asok_path:
2019-06-21T10:26:42.491 INFO:teuthology.orchestra.run.smithi141:>             assert(len(files) == 1)
2019-06-21T10:26:42.491 INFO:teuthology.orchestra.run.smithi141:>             return files[0]
2019-06-21T10:26:42.491 INFO:teuthology.orchestra.run.smithi141:>
2019-06-21T10:26:42.491 INFO:teuthology.orchestra.run.smithi141:>         for f in files:
2019-06-21T10:26:42.491 INFO:teuthology.orchestra.run.smithi141:>                 pid = re.match(".*\.(\d+)\.asok$", f).group(1)
2019-06-21T10:26:42.491 INFO:teuthology.orchestra.run.smithi141:>                 if os.path.exists("/proc/{0}".format(pid)):
2019-06-21T10:26:42.491 INFO:teuthology.orchestra.run.smithi141:>                         return f
2019-06-21T10:26:42.491 INFO:teuthology.orchestra.run.smithi141:>         raise RuntimeError("Client socket {0} not found".format(client_name))
2019-06-21T10:26:42.492 INFO:teuthology.orchestra.run.smithi141:>
2019-06-21T10:26:42.492 INFO:teuthology.orchestra.run.smithi141:> print find_socket("client.0")
2019-06-21T10:26:42.492 INFO:teuthology.orchestra.run.smithi141:> '
2019-06-21T10:26:42.529 INFO:teuthology.orchestra.run.smithi141.stdout:/var/run/ceph/ceph-client.0.25198.asok
2019-06-21T10:26:42.529 INFO:tasks.cephfs.fuse_mount:Found client admin socket at /var/run/ceph/ceph-client.0.25198.asok
2019-06-21T10:26:42.529 INFO:teuthology.orchestra.run:Running command with timeout 900
2019-06-21T10:26:42.530 INFO:teuthology.orchestra.run.smithi141:Running:
2019-06-21T10:26:42.530 INFO:teuthology.orchestra.run.smithi141:> sudo ceph --admin-daemon /var/run/ceph/ceph-client.0.25198.asok status
````

/ceph/teuthology-archive/pdonnell-2019-06-21_01:37:24-fs-wip-pdonnell-testing-20190620.220400-distro-basic-smithi/4052507/teuthology.log

The problem is that the `send_signal` method calls `get_client_pid` which uses the ceph-fuse admin socket to lookup the pid. Because the client is stopped (SIGSTOP), it cannot respond to the `status` asok command.

I recommend memoizing the `get_client_pid` call in `FuseMount` so that we don't unnecessarily run the `status` command multiple times needlessly. That nicely sidesteps the issue.

Reply:
@batrick Not just fuse daemon but even [`run_shell`](https://github.com/ceph/ceph/blob/master/qa/tasks/cephfs/mount.py#L185) won't work since the client process is stopped. I had to circumvent client and talk to vstart_runner.py directly. Locally, this test passed for me -

```
ceph-fuse[17731]: fuse finished with error 0 and tester_r 1
2019-07-02 22:17:59,070.070 INFO:__main__:Running ['rm', '-rf', '/tmp/tmpX7p31A/mnt.1']
2019-07-02 22:17:59,084.084 INFO:__main__:test_dont_mark_unresponsive_client_stale (tasks.cephfs.test_client_recovery.TestClientRecovery) ... ok
2019-07-02 22:17:59,084.084 INFO:__main__:Stopped test: test_dont_mark_unresponsive_client_stale (tasks.cephfs.test_client_recovery.TestClientRecovery) in 84.686263s
2019-07-02 22:17:59,084.084 INFO:__main__:
2019-07-02 22:17:59,084.084 INFO:__main__:----------------------------------------------------------------------
2019-07-02 22:17:59,084.084 INFO:__main__:Ran 1 test in 84.687s
2019-07-02 22:17:59,084.084 INFO:__main__:
2019-07-02 22:17:59,084.084 INFO:__main__:OK
```

 ...but teuthology might require some modification for this.


This also explains why tests on PR #28692 failed when they were passing locally.

Reply:
`run_shell` should work even if the ceph-fuse process is stopped.

Reply:
I remember `run_shell` had hanged right after command arguments were passed to subprocess[[1]](https://github.com/ceph/ceph/blob/master/qa/tasks/vstart_runner.py#L328). I guess the reason is that `cwd` is set to mountpoint[[2]](https://github.com/ceph/ceph/blob/master/qa/tasks/vstart_runner.py#L463)